src/foam/core/EmptyX.java

```diff
@@ -64,51 +56,128 @@ public Object getInstanceOf(Object value, Class type) {
   }
 }
 
-
-/** Default implementation of X interface. Stores one key-value binding. **/
-class XI
+abstract class AbstractXI
   extends AbstractX
+  implements Cloneable
 {
-  final X      parent_;
-  final Object key_;
-  final Object value_;
+  protected X   leftChild_;
+  protected X   rightChild_;
+  
+  protected abstract Object getKey();
+  protected X getLeftChild() { return leftChild_; }
```

Comment:
Why do these methods need to be in AbstractX? Couldn't we have an AbstractTreeX or AbstractXI which XI and FactoryXI extend? This way people can make other X implementations which don't follow this structure.

Reply:
??, It is in AbstractXI already